cmd/apprepository-controller/server/controller.go

```diff
@@ -386,504 +434,100 @@ func (c *Controller) ensureCronJob(apprepo *apprepov1alpha1.AppRepository) (meta
-}
-
-// truncateAndHashString truncates the string to a max length and hashes the rest of it
-// Ex: truncateAndHashString(aaaaaaaaaaaaaaaaaaaaaaaaaa,12) becomes "a-2067663226"
-func truncateAndHashString(name string, length int) string {
-	if len(name) > length {
-		if length < 11 {
-			return name[:length]
-		}
-		log.Warningf("Name %q exceeds %d characters (got %d)", name, length, len(name))
-		// max length chars, minus 10 chars (the adler32 hash returns up to 10 digits), minus 1 for the '-'
-		splitPoint := length - 11
-		part1 := name[:splitPoint]
-		part2 := name[splitPoint:]
-		hashedPart2 := fmt.Sprint(adler32.Checksum([]byte(part2)))
-		name = fmt.Sprintf("%s-%s", part1, hashedPart2)
-	}
-	return name
-}
-
-// apprepoSyncJobArgs returns a list of args for the sync container
-func apprepoSyncJobArgs(apprepo *apprepov1alpha1.AppRepository, config Config) []string {
-	args := append([]string{"sync"}, dbFlags(config)...)
-
-	if config.UserAgentComment != "" {
-		args = append(args, "--user-agent-comment="+config.UserAgentComment)
-	}
-
-	args = append(args, "--global-repos-namespace="+config.GlobalPackagingNamespace)
-	args = append(args, "--namespace="+apprepo.GetNamespace(), apprepo.GetName(), apprepo.Spec.URL, apprepo.Spec.Type)
-
-	if len(apprepo.Spec.OCIRepositories) > 0 {
-		args = append(args, "--oci-repositories", strings.Join(apprepo.Spec.OCIRepositories, ","))
+	if apprepo, ok = obj.(*apprepov1alpha1.AppRepository); !ok {
+		runtime.HandleError(fmt.Errorf("Error decoding object, invalid type"))
+		return
 	}
 
-	if apprepo.Spec.TLSInsecureSkipVerify {
-		args = append(args, "--tls-insecure-skip-verify")
+	// if the object is not being deleted (ie, deletionTimestamp==0)
+	if apprepo.GetDeletionTimestamp().IsZero() && !containsFinalizer(apprepo, finalizerName) {
+		// check if it contains the finalizer, if not, add it and update the object
+		log.Errorf("The AppRepository %q should be deleted, but doesn't have any finalizers. You might want to perform a manual clean-up", apprepo.GetName())
 	}
 
-	if apprepo.Spec.PassCredentials {
-		args = append(args, "--pass-credentials")
-	}
+	// if the object is being deleted and contains a finalizer
+	if !apprepo.GetDeletionTimestamp().IsZero() && containsFinalizer(apprepo, finalizerName) {
+		// if the object is being deleted and contains a finalizer and the event is not a deletion event,
+		// then handle the finalizer-derived clean-up tasks and remove the finalizer
+		if !shouldDelete {
+			log.Infof("Starting the clean-up tasks derived from the finalizer of the AppRepository %q", apprepo.GetName())
 
-	if apprepo.Spec.FilterRule.JQ != "" {
-		rulesJSON, err := json.Marshal(apprepo.Spec.FilterRule)
-		if err != nil {
-			log.Errorf("Unable to parse filter rules for %s: %v", apprepo.Name, err)
-		} else {
-			args = append(args, "--filter-rules", string(rulesJSON))
-		}
-	}
+			// start the clean-up
+			err = c.cleanUpAppRepo(apprepo, apprepo.GetNamespace(), apprepo.GetName())
```

Comment:
I know it sounds a bit hacky/tricky... but when testing IRL (and then confirmed in the k8s repo), a deletion is triggering also update events... so I ended up handling it that way. 
I did my best at explaining the code, but if you can think of a better solution, let me know :P